File: src/main/java/com/kamesuta/advquesting/data/CronParser.java
package com.kamesuta.advquesting.data;

import java.time.ZonedDateTime;
import java.time.temporal.ChronoUnit;
import java.util.BitSet;

/**
 * 簡易 cron パーサー。"分 時 日 月 曜日" の5フィールド形式をサポートする。
 * 対応: 数値, *, * /n, a-b, a,b
 * 曜日: 0=日曜, 1=月曜, ..., 6=土曜
 */
public class CronParser {

    private final BitSet minutes = new BitSet(60);  // 0-59
    private final BitSet hours   = new BitSet(24);  // 0-23
    private final BitSet days    = new BitSet(32);  // 1-31
    private final BitSet months  = new BitSet(13);  // 1-12
    private final BitSet dows    = new BitSet(7);   // 0-6 (日-土)

    private CronParser(String expr) {
        String[] fields = expr.trim().split("\\s+");
        if (fields.length != 5) throw new IllegalArgumentException("cron must have 5 fields: " + expr);
        parse(fields[0], minutes, 0, 59);
        parse(fields[1], hours,   0, 23);
        parse(fields[2], days,    1, 31);
        parse(fields[3], months,  1, 12);
        parse(fields[4], dows,    0, 6);
    }

    private static void parse(String field, BitSet bs, int min, int max) {
        if ("*".equals(field)) {
            bs.set(min, max + 1);
            return;
        }
        for (String part : field.split(",")) {
            if (part.startsWith("*/")) {
                int step = Integer.parseInt(part.substring(2));
                for (int i = min; i <= max; i += step) bs.set(i);
            } else if (part.contains("-")) {
                String[] range = part.split("-", 2);
                int lo = Integer.parseInt(range[0]);
                int hi = Integer.parseInt(range[1]);
                bs.set(lo, hi + 1);
            } else {
                bs.set(Integer.parseInt(part));
            }
        }
    }

    /** cron 式をパースする。不正な場合は null を返す */
    public static CronParser parse(String expr) {
        try { return new CronParser(expr); } catch (Exception e) { return null; }
    }

    /** now より前で最も直近の発火時刻を返す (分単位精度)。マッチしない場合 null */
    public static ZonedDateTime prevFire(String expr, ZonedDateTime now) {
        CronParser cp = parse(expr);
        if (cp == null) return null;
        // 現在分の1分前から過去24時間以内を探す (schedule クエストは最大1分遅延で検出)
        ZonedDateTime t = now.truncatedTo(ChronoUnit.MINUTES).minusMinutes(1);
        ZonedDateTime limit = now.minusHours(25);
        while (t.isAfter(limit)) {
            if (cp.matches(t)) return t;
            t = t.minusMinutes(1);
        }
        return null;
    }

    /** 次の発火時刻を返す（残り時間表示用・APIで使用）。最大366日先まで探す */
    public static ZonedDateTime nextFire(String expr, ZonedDateTime from) {
        CronParser cp = parse(expr);
        if (cp == null) return null;
        ZonedDateTime t = from.truncatedTo(ChronoUnit.MINUTES).plusMinutes(1);
        ZonedDateTime limit = from.plusDays(366);
        while (t.isBefore(limit)) {
            if (cp.matches(t)) return t;
            t = t.plusMinutes(1);
        }
        return null;
    }

    private boolean matches(ZonedDateTime t) {
        return minutes.get(t.getMinute())
            && hours.get(t.getHour())
            && days.get(t.getDayOfMonth())
            && months.get(t.getMonthValue())
            && dows.get(t.getDayOfWeek().getValue() % 7); // ISO: 1=月...7=日 → %7: 1=月,0=日
    }
}

File: web/src/components/editor/CronParser.ts
/**
 * 簡易 cron ユーティリティ (フロントエンド用)
 * "分 時 日 月 曜日" の5フィールド形式
 * nextFire: 次の発火時刻を返す (残り時間表示用)
 */

function parseBitSet(field: string, min: number, max: number): Set<number> {
  const s = new Set<number>()
  if (field === '*') {
    for (let i = min; i <= max; i++) s.add(i)
    return s
  }
  for (const part of field.split(',')) {
    if (part.startsWith('*/')) {
      const step = parseInt(part.slice(2), 10)
      for (let i = min; i <= max; i += step) s.add(i)
    } else if (part.includes('-')) {
      const [lo, hi] = part.split('-').map(Number)
      for (let i = lo; i <= hi; i++) s.add(i)
    } else {
      s.add(parseInt(part, 10))
    }
  }
  return s
}

function matches(d: Date, minutes: Set<number>, hours: Set<number>, days: Set<number>, months: Set<number>, dows: Set<number>): boolean {
  // JS: getDay() 0=日, 1=月…6=土 → 同じ
  return minutes.has(d.getMinutes())
    && hours.has(d.getHours())
    && days.has(d.getDate())
    && months.has(d.getMonth() + 1)
    && dows.has(d.getDay())
}

/** 次の発火時刻を返す。最大366日先まで探す。無効な式は null */
export function nextFire(expr: string, from: Date = new Date()): Date | null {
  const fields = expr.trim().split(/\s+/)
  if (fields.length !== 5) return null
  try {
    const minutes = parseBitSet(fields[0], 0, 59)
    const hours   = parseBitSet(fields[1], 0, 23)
    const days    = parseBitSet(fields[2], 1, 31)
    const months  = parseBitSet(fields[3], 1, 12)
    const dows    = parseBitSet(fields[4], 0, 6)

    // 現在分の次の分から探す
    const t = new Date(from)
    t.setSeconds(0, 0)
    t.setMinutes(t.getMinutes() + 1)

    const limit = new Date(from)
    limit.setDate(limit.getDate() + 366)

    while (t < limit) {
      if (matches(t, minutes, hours, days, months, dows)) return new Date(t)
      t.setMinutes(t.getMinutes() + 1)
    }
    return null
  } catch {
    return null
  }
}

/** 残り時間を "残りXh Ym" / "00:05 (6/20)" 形式に整形する */
export function formatCountdown(nextAt: Date): string {
  const now = new Date()
  const diffMs = nextAt.getTime() - now.getTime()
  if (diffMs <= 0) return '復活待機中'

  const totalMin = Math.floor(diffMs / 60000)
  const h = Math.floor(totalMin / 60)
  const m = totalMin % 60

  // 24時間未満は時分表示 (HH:MM形式 + 日をまたぐなら日付)
  if (h < 24) {
    const hh = String(h).padStart(2, '0')
    const mm = String(m).padStart(2, '0')
    const sameDay = nextAt.getDate() === now.getDate()
    if (!sameDay) {
      const mo = nextAt.getMonth() + 1
      const d = nextAt.getDate()
      return `${hh}:${mm} (${mo}/${d})`
    }
    return `${hh}:${mm}`
  }

  // 24時間以上は日時表示
  const days = Math.floor(h / 24)
  const remH = h % 24
  if (remH === 0) return `残り${days}d`
  return `残り${days}d ${remH}h`
}

/** クールダウン残り時間を返す (cooldownHours 後に復活) */
export function cooldownNextFire(completedAt: string, cooldownHours: number): Date {
  const d = new Date(completedAt)
  d.setTime(d.getTime() + cooldownHours * 3600000)
  return d
}
